### ISSUE TITLE
Proposal to modify `--quick` and/or add `--veryquick` option

### ISSUE THREAD

**Issue by donmelton** [donmelton/video_transcoding#108]
# Proposal to modify `--quick` and/or add `--veryquick` option

Earlier this month, I significantly increased the the speed and quality of the original `--quick` option. Encoding is now 70-80% faster than the default setting with no easily perceptible loss in video quality.

But I can't seem to leave well enough alone. :)

I now have a formula to increase encoding speed by 90-125% over the default setting. And _still_ with no easily perceptible loss in video quality.

You can try this proposed implementation yourself. Don't use `--quick` since that will invoke the current implementation. Instead add five new `--encoder-option` (or `-E`) arguments to your command line like this:

```
transcode-video -E partitions=none -E ref=1 -E bframes=1 -E rc-lookahead=30 -E me=dia "/path/to/Movie.mkv"
```

These five x264 video encoder settings are listed in order of their impact on performance with `partitions=none` being first since it gives the biggest boost.

The two new additions over the current `--quick` implementation are `bframes=1` and `me=dia`. It's `bframes=1` which adds the biggest new boost and demonstrates the cost for all that additional speed. Like `ref=1`, also part of the current `--quick` implementation, the `bframes=1` setting reduces the efficiency of the x264 encoder, meaning it takes more bits to encode the same quality.

But since my special ratecontrol is constrained by both target bitrate _and_ minimum quality, `bframes=1` doesn't reduce quality very much at all. At least not in a way that's easily perceptible. Instead, output bitrate increases somewhat. And if a video would exceed the target bitrate a little with the default setting, it will now exceed it a little more. And if it would exceed it a lot, it will now be a lot more.

Normally you would see less than a 1% increase in bitrate. Sometimes more than 1%. The worst case scenario might be a 10% increase. But, so far, I've only seen one movie go over 5%. So maybe the slightly larger size is not something to worry about.

One curiosity with the new formula is the large range of the possible speed increase: 90-125%. This is because some Blu-ray Discs contain video in VC-1 format instead of H.264 or MPEG-2. And HandBrake takes significantly longer to decode VC-1 format than the others.

If I were to exclude input files with VC-1 format video, the speed increase would be 110-125%. Now we can all be annoyed at Microsoft for inventing VC-1 and Warner Bros. for using it on their discs for so long. :)

The big question is: should this new formula replace the current `--quick` option or should I add a new `--veryquick` option to implement it?

I don't usually like to add new options but I'm concerned that some users might object to the possibility of slightly larger output bitrates with the new formula. And the new formula may seem like lower quality to those users, especially for DVD-sized content. Beauty is, after all, in the eye of the beholder.

BTW, if I do add a `--veryquick` option then I'll likely modify `--quick` anyway to remove the `mixed-refs=0` setting it currently uses. It turns out that setting is unnecessary when the `ref=1` setting is also applied.

Anyway, let me know what you think. Soon. Because I'm liable to make the change shortly after the first of the new year.

Thanks.

**Comment by RodBrown1988** [comment]
I'm going to err on the side of less options: make it the the new Quick. Unless the file size was dramatically larger, the trade off for simplicity and speed makes more sense to me.

**Comment by donmelton** [comment]
@RodBrown1988 Thanks for the feedback! Let me know if you try the formula.

**Comment by hughker** [comment]
All in for `--veryquick` but also okay with changing `--quick` to be these
new settings. I'm happy with both!!

Winston Hughes
https://winston.me
+1.313.598.5985

Sent from my iPhone

On Dec 26, 2016, at 17:57, Don Melton <notifications@github.com> wrote:

@RodBrown1988 <https://github.com/RodBrown1988> Thanks for the feedback!
Let me know if you try the formula.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/donmelton/video_transcoding/issues/108#issuecomment-269247371>,
or mute the thread

**Comment by hughker** [comment]
@donmelton I'm all in for `--veryquick` but also okay with changing `--quick` to be these
new settings. I'm happy with both!!

**Comment by RodBrown1988** [comment]
@donmelton after trying this a few times, I'd actually prefer a separate --veryquick option. --quick tends to produce significantly smaller file sizes (1.8GB vs 2.3GB for a 4:3 HD bluray copy), which is far more difference than I predicted. I think a --veryquick option makes more sense in this case. I personally use these transcodes for loading onto iOS devices, where size differences like that can quickly add up.

**Comment by donmelton** [comment]
@RodBrown1988 Yeah, I'm leaning toward two separate options now myself. Mostly because I don't want to lose that extra little bit of quality that `--quick` gives you. It's a nice middle-ground speed, too.

Thanks for testing and posting more feedback!

**Comment by donmelton** [comment]
@RodBrown1988 @hughker Here's how I'll describe the two options in the built-in help:

```
    --quick         increase encoding speed by 70-80%
                      with no easily perceptible loss in video quality
                      (avoids quality problems associated with x264 presets)
    --veryquick     increase encoding speed by 90-125%
                      with no easily perceptible loss in video quality
                      (avoids quality problems associated with x264 presets)
                      (unlike `--quick`, output size is larger than default)
```

The description for `--quick` won't change although, as I mentioned earlier, its implementation will very slightly.

Does this make sense?

**Comment by RodBrown1988** [comment]
That's great. Did you want to mention that the quality may be slightly lower?

**Comment by DaveHamilton** [comment]
Yeah, given the size delta I think the two options makes the most sense.

**Comment by donmelton** [comment]
@RodBrown1988 Yeah, I probably should mention the slightly lower quality. It's really hard to detect because it doesn't manifest itself with artifacts, color banding or as blockiness. It's just... not quite as good in a way that's both hard to see and hard to describe.

How about I change the line that says, "with no easily perceptible loss in video quality" to "with _little_ easily perceptible loss in video quality?"

**Comment by donmelton** [comment]
@DaveHamilton After I thought about it overnight and then seeing your and everyone else's feedback this morning, I'm planning on two options now. Like minds. :) Thanks!

**Comment by donmelton** [comment]
@RodBrown1988 @hughker @DaveHamilton OK, here's the revised text for the built-in help:

```
    --quick         increase encoding speed by 70-80%
                      with no easily perceptible loss in video quality
                      (avoids quality problems associated with x264 presets)
    --veryquick     increase encoding speed by 90-125%
                      with little easily perceptible loss in video quality
                      (unlike `--quick`, output size is larger than default)
```

While the `--veryquick` option also avoids quality problems associated with x264 presets, I didn't think it bore repeating there.

For those of you, like @DaveHamilton, who tested my earlier "turbo" formula last week, you'll notice that `--veryquick` is not quite as fast. But it's also considerably better quality in some cases than that first attempt at a speedup. That's because I didn't change x264's default `subme=7` setting to `subme=6` with this new option.

What the Sheol is `subme`? That's short for "sub-pixel motion estimation and mode decision." And it really controls how x264 deals with rate distortion. Which, of course, takes time. And that makes it a tempting setting to change in order to increase performance.

Lower `subme` too much, like x264 does with the `--faster`, `--veryfast`, etc. presets, and you'll start to see artifacts, blockiness and color banding quite often. But even lowering it a little will cause some of those problems in a few places. Worse, it will lower the amount of detail you're able to retain.

Another setting I didn't use from that earlier "turbo" formula is `mbtree=0`. This completely disables the macroblock-based ratecontrol algorithm in x264, essentially rendering all B-frames at the same level of quality, among other behavior changes it invokes.

Weirdly, using `mbtree=0` improves quality in some cases, but it also causes just as many if not more quality problems in others. Worse, like using `bframes=1`, it makes for larger output bitrates. So it's not a good idea to combine those two settings or file sizes really get big.

TL;DR: The `--veryquick` option still produces remarkably high quality output. :)

**Comment by Eltito-Houston** [comment]
Don as per our Twitter exchange this morning I am attaching the relevant files on my Transcoding Exercise with the STD algoritm the "--Quick" option and the new 5 Parameters request ("--VeryQuick"). The speed improvements are remarkable but the only thing that call my attention is that the resultant file sizes are Identical

[Don Melton Transcode Script.docx](https://github.com/donmelton/video_transcoding/files/676155/Don.Melton.Transcode.Script.docx)
[LogFiles.zip](https://github.com/donmelton/video_transcoding/files/676156/LogFiles.zip)

**Comment by RodBrown1988** [comment]
I've also done quite a bit of transcoding with both and found that my initial file size comparisons ended up being edge cases, it seems, with most others actually ending up being identical. I concur with @Eltito-Houston that for most cases, the file sizes appear comparable.

With this in mind the issue for --quick vs --veryquick becomes more about the quality trade-offs. I'm not seeing much difference at all. That said, I'm currently not transcoding stuff that would push it to its limit.

@donmelton I suppose this comes down to your feeling on this. I find them very close, except this new formula ends up simply significantly faster.

**Comment by Eltito-Houston** [comment]
The "--VeryQuick" option result in a remarkable 2Hrs - 28 Min shorter encoding time in My 2006 MacPro system. Not sure in a modern system how much it could be

**Comment by RodBrown1988** [comment]
I'm finding 43 minute episodes of 4:3 cropped TV shows at 1080p drop from 17 mins to 12 mins. That's on a retina 5K iMac late 2014, 4GHz. Definitely an improvement.

**Comment by donmelton** [comment]
@Eltito-Houston Sorry I took so long to look at your data. I had a three-hour drive this morning followed by a tasty lunch. :)

First, you need to redo "Task#3", the "veryquick" test, because it appears one of your encoder arguments are wrong. You must have specified `-E rc-1lookahead=10` on the command line, which is not a valid option. The good news is that this probably caused only a minor 3 to 5% slowdown because your other arguments were correct.

The proper argument is `-E rc-lookahead=30`. Just go back to the top of this thread and make sure you copy and paste all five arguments the next time. That's what I always do because I'm forever making typos. :)

Other than that slight slowdown, the speed and bitrate data is what I would expect from all three tests.

**Comment by Eltito-Houston** [comment]
So in my case a 128 Minutes 1080p 2:35 widescreen Movie result in a reduction from 240 Minutes to 120.6 Minutes, there should be a Math explanation for sure on the time reduction

**Comment by donmelton** [comment]
@RodBrown1988 Yeah, it's hard call to make! On the one hand, typical file sizes aren't growing all that much with the proposed `--veryquick` formula and its quality seems very nearly on par with `--quick`.

But using `query-handbrake-log` I can see that the average P-frame quantizer is slightly higher for `--veryquick` output compared to that from `--quick`. And higher values are, of course, worse.

I'll probably wrestle with this until after the new year. :)

**Comment by Eltito-Houston** [comment]
Don, I will repeat test #3 and reply back with the updated results. Thanks for the correction

**Comment by donmelton** [comment]
@Eltito-Houston Sounds like what I predicted, a nearly 100% reduction in time. :)

**Comment by Eltito-Houston** [comment]
Don:

I have done Task#3 again (Added as task #4). The argument "E rc-lookahead=30" doesn´t seen to change the results. In fact the new test took 7 minutes longer. Now you have 4 logs files to play with.

Cheers

[Logs File.zip](https://github.com/donmelton/video_transcoding/files/676524/Logs.File.zip)
[Don Melton Transcode Script.docx](https://github.com/donmelton/video_transcoding/files/676525/Don.Melton.Transcode.Script.docx)

**Comment by donmelton** [comment]
@Eltito-Houston Thanks for re-testing that so quickly!

However, it seems you used exactly the same command line options as before. Look at line number 269 in "Task#4 2016.mp4.log" and you'll find `partitions=none:ref=1:bframes=1:rc-1lookahead=30:me=dia`. This line shows exactly what you used when invoking `transcode-video`.

There is no x264 setting spelled `rc-1lookahead`. It's spelled `rc-lookahead`, without the `1` in the middle.  I'm not sure why you keep inserting the number `1` between `rc-` and `lookahead` there, but it will be ignored by the encoder. :)

As to why this run is slower, I have no idea. Is your computer "quiet" when running? By that I mean other processes should be not be sucking time away from the encoder. If your computer isn't "quiet" when you run the tests, then you won't get reliable and repeatable data.

I would recommend quitting as many applications as possible before running your tests, especially programs like Web browsers.

**Comment by RodBrown1988** [comment]
... oh the irony of @donmelton telling people to quit web browsers...

I'm sorry, I couldn't help but comment...

**Comment by donmelton** [comment]
@RodBrown1988 Oh, the irony wasn't lost on me. :)

**Comment by Eltito-Houston** [comment]
Guys, I am sorry will re-test again

**Comment by donmelton** [comment]
@Eltito-Houston No worries. This kind of thing is hard. You wouldn't believe the odd mishaps I've had during performance testing. It's enough to make you crazy at times.

**Comment by Eltito-Houston** [comment]
Computer is running Quiet and I haven't touch anything since I start doing all the testing. Running now with the rc-lookahead=30, I didn't saw that number "1" probably I need new glasses ):

**Comment by Eltito-Houston** [comment]
Ok.

I believe now the command is properly imput and Not visible changes in Speed or Size.

Still the "--VeryQuick" Sholw to be sligthly faster than the "--Quick"

Cheers


[Don Melton Transcode Script.docx](https://github.com/donmelton/video_transcoding/files/677107/Don.Melton.Transcode.Script.docx)
[Task#4.log.zip](https://github.com/donmelton/video_transcoding/files/677106/Task.4.log.zip)

**Comment by donmelton** [comment]
@Eltito-Houston Thanks! Yeah that looks right and the numbers are more like what I would expect.

**Comment by Eltito-Houston** [comment]
No Problems Don. Considering the results I am setting myself for the "--VeryQuick" Algorithm going forward. Let me know any other testing you want to do. Cheers

**Comment by ericcardenas** [comment]
Go for it Don. 
I welcome the change. 
All for new `--veryquick’ and for the small change in `--quick’.

**Comment by donmelton** [comment]
@ericcardenas Cool. It will likely be checked in and released on Monday or Tuesday.

**Comment by donmelton** [comment]
@RodBrown1988 @hughker @DaveHamilton @Eltito-Houston @ericcardenas OK, folks, the slightly modified `--quick` option and the brand new `--veryquick` option are now available in lucky number version 0.13.0. Just `gem update video_transcoding` to get it.

Thanks again for all your feedback and testing!

**Comment by Eltito-Houston** [comment]
Don:

this is the latest log file from the "File" we have been discussing on the past week. I "Transcode" using version 0.13 and to test the speed on my Win10 - i7 laptop (Better than my Old MacPro ): ) in any case it did faster but also spit out a considerable 1.6 GB smaller file compare to the previous results!!

Any comments on your side?



[Don_Melton Transcode Videos.docx](https://github.com/donmelton/video_transcoding/files/688749/Don_Melton.Transcode.Videos.docx)
[The Accountant6 2016.txt.zip](https://github.com/donmelton/video_transcoding/files/688748/The.Accountant6.2016.txt.zip)

**Comment by Eltito-Houston** [comment]
I am sorry. Posted the wrong Word document
[Don Melton Transcode Script.docx](https://github.com/donmelton/video_transcoding/files/688760/Don.Melton.Transcode.Script.docx)

**Comment by donmelton** [comment]
@Eltito-Houston I can't really tell without looking at the two `.log` files side by side. Your finale video bitrate here was 5971.34 kbps. Was the other file somehow larger?

**Comment by Eltito-Houston** [comment]
Don, it is in Task#4.log.zip posted above

**Comment by donmelton** [comment]
@Eltito-Houston Thanks. OK, the final video bitrate on the first file from the Mac is 5869.74 kbps so it's actually _smaller_ than the file from Windows.

I'm not sure how you're doing your size measurements, but there's clearly something going on wrong there. :)

**Comment by Eltito-Houston** [comment]
Don, repeat the test Yesterday on the macpro with the same results. I have posted some screen capture and the logs files. Very interesting stuff here when you run Transcode in Windows and mac OSX with the same file. Is there somebody else running it in Both Systems?


![side by side logs](https://cloud.githubusercontent.com/assets/24815102/21719576/1909b8c0-d3e4-11e6-90bb-74e4440261fd.jpg)
![finder_capture size](https://cloud.githubusercontent.com/assets/24815102/21719603/33f6e9b4-d3e4-11e6-9df2-b3f29b6ef9e8.jpg)
[The Accountant6 2016.txt.zip](https://github.com/donmelton/video_transcoding/files/690040/The.Accountant6.2016.txt.zip)

**Comment by donmelton** [comment]
@Eltito-Houston  I've had my coffee so I can look at that screenshot now... :)

Thanks for doing a real side-by-side comparison! Wow, that's going the extra mile, sir.

OK, there's something weird going on here. Although the output video (and audio) bitrates are very close to the same values, I just noticed that the output byte sizes of all three of the streams are much larger on your Mac Pro. Which... makes no sense to me why that should ever happen.

Honestly, I have no explanation for this. Yet.

It looks like I may have to setup my own Windows 10 machine for transcoding to figure this out.

Thanks for bringing this to my attention!!! Seriously. In fact, you may want to open a new GitHub issue with all this data so we can track the mystery separately.

**Comment by Eltito-Houston** [comment]
Don, can you take a look on the log lines "Sync: got XXXXXX Frames, Expected XXXXXX for both files and also de Line "Mux: Track0 XXXXXX Frames" for some reason the MacPro is showing "Twice" of the reported Sync Value compare to the Windows System also the FIFO Bynary is twice 1024 Vs 2048 vs 4096

**Comment by donmelton** [comment]
@Eltito-Houston Nice catch. Yes, the "got" and "expected" frames could explain it all. Here's the line perfectly normal line on your Mac:

```
sync: got 183999 frames, 183999 expected
```

... and here's the totally weird line on your PC:

```
sync: got 136437 frames, 183999 expected
```

That tells me the PC version only got 74% of the frames. So, my guess is that it is somehow truncated. Which would explain its smaller size.

What does VLC, mediainfo, or other tools say the duration of the two files are? Do they match? If you play the PC file with VLC, can you actually see the whole thing?

**Comment by Eltito-Houston** [comment]
Don, You are 100% correct and the PC file is truncated. It does play in VLC but is missing the last 15 Minutes. I try again with the same result and also with a different movie file and the problem persist.

I will need more time to troubleshoot why Transcoding is truncating files in Windows but need another PC to discard the Hardware portion.

One more thing isssue isn't related to "--Veryquick" option the "--Quick" parameter have similar behavior.

Cheers

**Comment by donmelton** [comment]
@Eltito-Houston Wow. And thanks for doing so much testing!

OK, good to know it has nothing to do with `--veryquick` or `--quick`. Now we have to figure out why your files are getting truncated on Windows.

HandBrake had a bug like that years ago but it was never that severe. I'll peruse their bug database and changelog to see if anyone else is having a problem like this. Stay tuned...

**Comment by donmelton** [comment]
@Eltito-Houston OK, I haven't found any open issues or source code changes in the various HandBrake databases that describe the problem you're seeing on Windows. So, my advice is to try and reproduce the problem with `HandBrakeCLI` only and open a new issue on their GitHub project.

The easiest way to take `transcode-video` out of the steps needed to reproduce your problem is to first run the same command line with the `--dry-run` option. That will print out the actual command that` transcode-video` is sending to `HandBrakeCLI`.

You can then take that `HandBrakeCLI`-only command line and start eliminating other options until you have the minimum needed to reproduce the truncated file problem.

Good luck and let me know if you need additional help with this.
